pymtl3/passes/sverilog/import_/verilator_wrapper.py.template

```diff
@@ -0,0 +1,142 @@
+  nbits = wire.nbits
+  if nbits <= 64:
+    return to_string( signal[0], nbits )
+  else:
+    ret = ""
+    num_elements = (nbits-1)/32+1
+    for idx in range(num_elements):
+      _nbits = 32 if idx != num_elements-1 else nbits%32
+      ret = to_string( signal[idx], _nbits )[2:] + ret
+    return "0x" + ret
+
+#-------------------------------------------------------------------------
+# {component_name}
+#-------------------------------------------------------------------------
+class {component_name}( Component ):
+  id_ = 0
+
+  def __init__( s ):
+    s._finalization_count = 0
+
+    # initialize FFI, define the exposed interface
+    s.ffi = FFI()
+    s.ffi.cdef("""
+      typedef struct {{
+
+        // Exposed port interface
+{port_cdefs}
+
+        // Verilator model
+        void * model;
+
+      }} V{component_name}_t;
+
+      V{component_name}_t * create_model();
+      void destroy_model( V{component_name}_t *);
+      void eval( V{component_name}_t * );
+
+    """)
+
+    # Print the modification time stamp of the shared lib
+    # print 'Modification time of {{}}: {{}}'.format(
+      # '{lib_file}', os.path.getmtime( './{lib_file}' ) )
+
+    # Import the shared library containing the model. We defer
+    # construction to the elaborate_logic function to allow the user to
+    # set the vcd_file.
+    s._ffi_inst = s.ffi.dlopen('./{lib_file}')
+
+    # increment instance count
+    {component_name}.id_ += 1
+
+  def finalize( s ):
+    """Finalize the imported component.
+    
+    This method closes the shared library opened through CFFI. If an imported
+    component is not finalized explicitly (i.e. if you rely on GC to collect a
+    no longer used imported component), importing a component with the same
+    name before all previous imported components are GCed might lead to
+    confusing behaviors. This is because once opened, the shared lib
+    is cached by the OS until the OS reference counter for this lib reaches
+    0 (you can decrement the reference counter by calling `dl_close()` syscall).
+
+    Fortunately real designs tend to always have the same shared lib corresponding
+    to the components with the same name. If you are doing translation testing and
+    use the same component class name even if they refer to different designs,
+    you might need to call `imported_object.finalize()` at the end of each test
+    to ensure correct behaviors.
+    """
+    assert s._finalization_count == 0,\
+      'Imported component can only be finalized once!'
+    s._finalization_count += 1
+    s._ffi_inst.destroy_model( s._ffi_m )
+    s.ffi.dlclose( s._ffi_inst )
+    s.ffi = None
+    s._ffi_inst = None
+
+  def __del__( s ):
+    if s._finalization_count == 0:
+      s._finalization_count += 1
+      s._ffi_inst.destroy_model( s._ffi_m )
+      s.ffi.dlclose( s._ffi_inst )
+      s.ffi = None
+      s._ffi_inst = None
+
+  def construct( s ):
+    # Construct the model
+    s._ffi_m = s._ffi_inst.create_model()
```

Comment:
we can do ffi_m = s._ffi_m and use ffi_m in update blocks to remove some redundant bytecode for cpython

Reply:
Done.